### PR TITLE
BUG: Check types in Index.__contains__ (#22085)

### LINKED ISSUE #22085: Checking that a pandas.Series.index contains a value
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
import numpy as np
import pandas as pd

some_numbers = np.random.randint(0,4,size=10)
print(some_numbers)
s = pd.Series(some_numbers)
gb = s.groupby(s).size() / len(s)
print(gb)
1.3 in gb
```
#### Problem description
I reported it [here](https://stackoverflow.com/questions/51563433/checking-that-a-pandas-series-index-contains-a-value). The output of the last line is True instead of being False.

#### Expected Output
False

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.2.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.3
pytest: 3.2.1
pip: 9.0.1
setuptools: 40.0.0
Cython: 0.26.1
numpy: 1.15.0
scipy: 0.19.1
pyarrow: None
xarray: None
IPython: 6.1.0
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.2.2
openpyxl: 2.4.8
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 0.9.8
lxml: 3.8.0
bs4: 4.6.0
html5lib: 0.999999999
sqlalchemy: 1.1.13
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by yeojin-dev**
- [x] closes #22085
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I added is_float, is_integer_dtype in Index.\_\_contains\_\_.
If key is float and dtype of Index object is integer, return False.

### CODE REVIEW

**Comment by jreback** [comment]
note that this duplicates #22366 

need tests & a whatsnew entry

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -653,6 +653,7 @@ Indexing
 - Fixed ``DataFrame[np.nan]`` when columns are non-unique (:issue:`21428`)
 - Bug when indexing :class:`DatetimeIndex` with nanosecond resolution dates and timezones (:issue:`11679`)
 - Bug where indexing with a Numpy array containing negative values would mutate the indexer (:issue:`21867`)
+- Bug in :func:`Index.__contains__` that returns True just in case val is float and dtype is integer, because float casted to integer (:issue:`22085`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: something more along the line of 
```Bug in 'scalar in Index' if scalar is a float while the Index is of integer dtype.```

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -1949,7 +1949,11 @@ def __nonzero__(self):
     def __contains__(self, key):
         hash(key)
         try:
-            return key in self._engine
+            if is_float(key) and is_integer_dtype(self.dtype) and\
```
Comment (jreback) [inline pandas/core/indexes/base.py]: don't use ``\`` for line continuation, instead use parens

File: pandas/tests/indexes/test_base.py [hunk 3]
```diff
@@ -2483,6 +2483,14 @@ def test_comparison_tzawareness_compat(self, op):
         # TODO: implement _assert_tzawareness_compat for the reverse
         # comparison with the Series on the left-hand side
 
+    def test_contains_with_float_val(self):
+        # GH#22085
+        index = pd.Index([0, 1, 2, 3])
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: can you also add a Float index and test the same cases

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by yeojin-dev** [comment]
@jreback I fix them. Please check. 😊

**Comment by yeojin-dev** [comment]
@jreback Could you please give me your feedback on my last commit?

File: doc/source/whatsnew/v0.24.0.txt [hunk 4]
```diff
@@ -658,6 +658,7 @@ Indexing
 - Bug when indexing :class:`DatetimeIndex` with nanosecond resolution dates and timezones (:issue:`11679`)
 - Bug where indexing with a Numpy array containing negative values would mutate the indexer (:issue:`21867`)
 - ``Float64Index.get_loc`` now raises ``KeyError`` when boolean key passed. (:issue:`19087`)
+- Bug in `scalar in Index` if scalar is a float while the Index is of integer dtype (:issue:`22085`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: use double-backticks on the ``Index`` (2nd use in sentence)

File: pandas/core/indexes/base.py [hunk 5]
```diff
@@ -1993,7 +1993,11 @@ def __nonzero__(self):
     def __contains__(self, key):
         hash(key)
         try:
-            return key in self._engine
+            if (is_float(key) and is_integer_dtype(self.dtype) and
+                    int(key) != key):
+                return False
+            else:
```
Comment (jreback) [inline pandas/core/indexes/base.py]: don't need the else here.

File: pandas/core/indexes/base.py [hunk 6]
```diff
@@ -1993,7 +1993,11 @@ def __nonzero__(self):
...
+                return key in self._engine
```
Comment (jreback) [inline pandas/core/indexes/base.py]: this is ok in python code, but actually can you see if you can do this in the cython code?

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt (hunk 4 above)
Comment (yeojin-dev) [reply doc/source/whatsnew/v0.24.0.txt]: Done.

File: pandas/core/indexes/base.py (hunk 5 above)
Comment (yeojin-dev) [reply pandas/core/indexes/base.py]: Done.

File: pandas/core/indexes/base.py (hunk 6 above)
Comment (yeojin-dev) [reply pandas/core/indexes/base.py]: Thanks. But I'm not good at Cython now. I'd like to try it but it'll take time.

**Comment by yeojin-dev** [comment]
@jreback My new commits has just passed a test in my Mac. But can't here, GitHub.

See the logs below.

1. Travis
```
ImportError: libgfortran.so.1: cannot open shared object file: No such file or directory
The command "ci/install_travis.sh" failed and exited with 1 during .
```

2. Circle
```
ImportError: libgfortran.so.1: cannot open shared object file: No such file or directory
```

I guess these errors are not caused by my commits. Please check.

**Comment by yeojin-dev** [comment]
@jreback I wrote the code in Cython. But my code has just passed the test, can't pass here. I can't understand... 😢

**Comment by jreback** [comment]
can you rebase

**Comment by yeojin-dev** [comment]
@jreback Sorry. I deleted the branch while rebasing it. Please see my new PR.

**Pull request closed without merging**